Common/Data/Market/Tick.cs

```diff
@@ -207,6 +208,145 @@ public Tick(Symbol symbol, string line, DateTime baseDate)
+        /// <param name="reader">The source stream reader</param>
+        /// <param name="date">Base date for the tick (ticks date is stored as int milliseconds since midnight)</param>
+        /// <param name="config">Subscription configuration object</param>
+        public Tick(SubscriptionDataConfig config, StreamReader reader, DateTime date)
+        {
+            try
+            {
+                DataType = MarketDataType.Tick;
+                Symbol = config.Symbol;
+
+                // Which security type is this data feed:
+                var scaleFactor = GetScaleFactor(config.SecurityType);
+
+                switch (config.SecurityType)
+                {
+                    case SecurityType.Equity:
+                        {
+                            TickType = config.TickType;
+                            Time = date.Date.AddMilliseconds((double)reader.GetDecimal()).ConvertTo(config.DataTimeZone, config.ExchangeTimeZone);
+
+                            bool pastLineEnd;
+                            if (TickType == TickType.Trade)
+                            {
+                                Value = reader.GetDecimal() / scaleFactor;
+                                Quantity = reader.GetDecimal(out pastLineEnd);
+                                if (!pastLineEnd)
+                                {
+                                    Exchange = reader.GetString();
+                                    SaleCondition = reader.GetString();
+                                    Suspicious = reader.GetInt32() == 1;
+                                }
+                            }
+                            else if (TickType == TickType.Quote)
+                            {
+                                BidPrice = reader.GetDecimal() / scaleFactor;
+                                BidSize = reader.GetDecimal();
+                                AskPrice = reader.GetDecimal() / scaleFactor;
+                                AskSize = reader.GetDecimal(out pastLineEnd);
+
+                                SetValue();
+
+                                if (!pastLineEnd)
+                                {
+                                    Exchange = reader.GetString();
+                                    SaleCondition = reader.GetString();
+                                    Suspicious = reader.GetInt32() == 1;
+                                }
+                            }
+                            else
+                            {
+                                throw new InvalidOperationException($"Tick(): Unexpected tick type {TickType}");
+                            }
+                            break;
+                        }
+
+                    case SecurityType.Forex:
+                    case SecurityType.Cfd:
+                        {
+                            TickType = TickType.Quote;
+                            var ticks = (long)(reader.GetDecimal() * TimeSpan.TicksPerMillisecond);
```

Comment:
These are int-milliseconds; we should review all to confirm using the correct type and using the AddMillisecond overload when appropriate